### DISCUSSION TITLE
Changes in V2 version

### DISCUSSION THREAD

**Discussion by Mister-Hope** [vuepress-theme-hope/vuepress-theme-hope#601] [Annoucement]
## Details

V2 will come together with _VuePress@v2_, with the power of _vite@v2_ _webpack@v5_ and _vue@v3_.

V2 will be a totally rewrite including:

- All the components will be rewrite with _composition api in vue3_
- All the styles will be mirgate to _sass_

Also we are going to make sure the rewrite one has a better performance!🚀 (no less than 25% increasement)

---

Below is a list of changes.

> Note: All plugin names are in PascalCase and without `vuepress-plugin-` prefix, you should change to `kebab-case` and add the prefix yourself to get the correct plugin name.

## AddThis💌 ![deprecated](https://img.shields.io/badge/-deprecated-yellow)

We are just making this plugin compatable with V2. This plugin will no longer under maintainment.

### Why❔

AdBlockers are blocking addThis link heavily, and in most cases this plugin won't work.

### Solution💡

We will provide a new share plugin in v2.

## Comment2💬

Renamed from `@mr-hope/vuepress-plugin-comment` to `vuepress-plugin-comment2`. ⚠

### Changes

- [x] Drop Valine support! ![removed](https://img.shields.io/badge/-removed-red)

  Valine is lack of maintainence and can leak your privacy. You should use [Waline](https://waline.js.org) instead.

- [x] Option `hint` (default `true`) for handling page info hints. ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Allow I18N Customize ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Muti categories support ![new](https://img.shields.io/badge/-new-brightgreen)

### Note

Vssue is currently being dropped because vssue has not support v2 yet. We will add it back after it supports v2.

## CopyCode2📋

Renamed from `@mr-hope/vuepress-plugin-copy-code` to `vuepress-plugin-copy-code2`. ⚠

### Changes

- [x] Option `delay` (default `500`) for handling the page swtiching transition. ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Option `pure` (default `false`) for handling fancy styles. ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Allow I18N Customize ![new](https://img.shields.io/badge/-new-brightgreen)

## Feed2

Renamed from `@mr-hope/vuepress-plugin-feed` to `vuepress-plugin-feed2`. ⚠

### Changes

- [x] Better log output. ![improve](https://img.shields.io/badge/-improved-blue)

- [x] Muti categories support ![new](https://img.shields.io/badge/-new-brightgreen)

### Changes

- [x] Auto inject cdn links for markdown demo ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Option `delay` (default `500`) for handling the page swtiching transition. ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Drop flowchart support ![removed](https://img.shields.io/badge/-removed-red)

  We are adding mermaid support, and mermaid support flowchart.

- [x] Move SlidePage layout from theme to the plugin

  This will allow other theme to provided a similar layout easily.

- [x] Allow I18N Title in details block ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Allow I18N Customize ![new](https://img.shields.io/badge/-new-brightgreen)

## LightGallery ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Build a photo preview plugin using [lightGallary](https://www.lightgalleryjs.com/)

## Pwa2

Renamed from `@mr-hope/vuepress-plugin-pwa` to `vuepress-plugin-pwa2`. ⚠

### Changes

- [x] Allow I18N Customize ![new](https://img.shields.io/badge/-new-brightgreen)

### Changes

- [x] Option `delay` (default `500`) for handling the page swtiching transition. ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Allow I18N Customize ![new](https://img.shields.io/badge/-new-brightgreen)

## ReadingTime2

Renamed from `@mr-hope/vuepress-plugin-reading-time` to `vuepress-plugin-reading-time2`. ⚠

- [x] Allow I18N Customize ![new](https://img.shields.io/badge/-new-brightgreen)

## SassPalette ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] A new palette handling config and palette across theme and plugins using sass

Options:

```ts
export interface PaletteOptions {
  /** identifier for palette */
  id: string;

  /**
   * user config file path, relative to source dir
   *
   * @default ".vuepress/styles/config.scss"
   */
  config?: string;

  /**
   * default config file path, should be absolute path
   *
   * This is the file you should use to provide default values with `!default`
   *
   * @default "vuepress-plugin-sass-palette/styles/default/config.scss"
   */
  defaultConfig?: string;

  /**
   * user palette file path, relative to source dir
   *
   * @default ".vuepress/styles/palette.scss"
   */
  palette?: string;

  /**
   * default palette file path, should be absolute path
   *
   * This is the file you should use to provide default values with `!default`
   *
   * @default "vuepress-plugin-sass-palette/styles/default/palette.scss"
   */

  defaultPalette?: string;

  /**
   * user style file path, relative to source dir
   *
   * @default ".vuepress/styles/palette.scss"
   */
  style?: string;
}
```

## SEO2

Renamed from `@mr-hope/vuepress-plugin-seo` to `vuepress-plugin-seo2`. ⚠

### Changes

- [x] JSON-LD support ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Better log output. ![improve](https://img.shields.io/badge/-improved-blue)

- [x] Muti categories support ![new](https://img.shields.io/badge/-new-brightgreen)

## Share2 ![new](https://img.shields.io/badge/-new-brightgreen)

- [ ] A new share plugin providing page share.

## SimpleBlog ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Build a own blog plugin to support blog with better performance

## Sitemap2

Renamed from `@mr-hope/vuepress-plugin-sitemap` to `vuepress-plugin-sitemap2`. ⚠

### Changes

- [x] Better log output. ![improve](https://img.shields.io/badge/-improved-blue)

- [x] Muti categories support ![new](https://img.shields.io/badge/-new-brightgreen)

## VuepressThemeHope

- [x] Pure Mode in theme and all plugins #438 ![new](https://img.shields.io/badge/-new-brightgreen)

- [ ] Dim non focused area #168 ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Code block theme config #189 ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Muti categories support ![new](https://img.shields.io/badge/-new-brightgreen)

- [x] Drop fullscreen feature, or make it a optional feature (disabled by default). ![removed](https://img.shields.io/badge/-removed-red)

  Most visitors do not use this feature.

- [x] Drop themecolor feature, or make it a optional feature (disabled by default). ![removed](https://img.shields.io/badge/-removed-red)

  Most visitors do not use this feature.

- [x] Allow I18N Customize ![new](https://img.shields.io/badge/-new-brightgreen)

**Comment by loctran016** [comment]
Hi there, your lightbox is great, but it will be better if you add [lightGallery](https://www.lightgalleryjs.com/) as lightGallery has even more features (and it have a Vue version, too)

**Reply by loctran016** [reply]
Is adding [typed.js](https://github.com/mattboldt/typed.js/) to the tagline to make it look even better acceptable? Btw, it would be nice if you could add [utterances](https://github.com/utterance/utterances) (It have more theme than vssue although it only support github)

**Reply by Mister-Hope** [reply]
Some feedback about lightGallery:

- Extra deps: See https://github.com/sachinchoolur/lightGallery/issues/1047

- It's mainly usage is wrapping a set of images, but our images are listed sperately in the whole docs content. Which I need extra codes and performance lost to make it compatable.

-  It's size is much more bigger than phowo-swipe. It's core function is probably the same size as the whole photoswipe, but it's even spliting lots of features in to plugins, e.g.: fullscreen, zoom, thumbnail. Which means it will be finally 1.5 - 2 times big comparing to photoswipe

- Nearly 0 benifit with photoswipe. Expect some improvements in UIs, 

   I don't think people will use rotate or tranverse function on the pics every offen. And we cannot write photographer, location and such info in markdown image alt. 

  Videos support is great, but the plugin is much bigger than I expected, and it's not a common need, so even I build this plugin, this will probably not enabled by default.

We think this project is awesome, mainly because the ui is awesome. So I think it's time to rethink whether I should build this plugin.

**Reply by Mister-Hope** [reply]
It will be dropped. It's a big time waste for me. I almost finish writing the plugin, then I found this:

![image](https://user-images.githubusercontent.com/33315834/121794108-4bfc6880-cc38-11eb-874d-62a8286ff0dc.png)

It's open souce but actually not free.

https://github.com/sachinchoolur/lightGallery/blob/master/LICENSE

See, you can only use it in GNU GPL license v3 projects for free, otherwise you must purchase a license. And that's very expensive.

My project will stay on MIT, that's unaccpetable for me.

Just a recommandation, please check if it's something I can use next time you recommand something to me 🤔

**Reply by Mister-Hope** [reply]
Update: see https://github.com/sachinchoolur/lightGallery/issues/1048

**Reply by loctran016** [reply]
I'm truly sorry about this inconvenience, I used to think this project is GNU GPL license v3 project. I'm reallyyyyyyy sorry for this!!!!!!! 😭

**Reply by Mister-Hope** [reply]
It's ok, jsut a recommandation, I should watch the license first.

**Comment by dimaslanjaka** [comment]


**Reply by dimaslanjaka** [reply]
![image](https://user-images.githubusercontent.com/12471057/120910487-864e8e80-c6a9-11eb-9fb2-6adec350540c.png)

**Reply by dimaslanjaka** [reply]
First checkout lint error

![image](https://user-images.githubusercontent.com/12471057/120911066-5f468b80-c6ae-11eb-9a58-0841e0cebcbc.png)

**Reply by Mister-Hope** [reply]
I will write a guide very soon.

**Reply by Mister-Hope** [reply]
Just to be short,

1. we are using commonjs(cjs) on node side, and es module(esm) on client side, so you can not mixed it.

2. The whole project is using `ts composite`, which you must build from root folder. The projects under packages will be built by the sequence they are requiring each other.

3. We are using prettier and type-checked @typescript-eslint/plugin, and use husky and lint-staged to add hooks to pre-commit, means you can only commit if the style and types are correct.

4. Do `yarn clean` when you have code changes in vuepress-shared package, that's a limitation due to type files imported by other projects

5. You should run `ts:watch` and `copy:watch` when you develop.

**Reply by Mister-Hope** [reply]
I will add a much more detailed  one later in docs.

**Reply by Mister-Hope** [reply]
Contribution guide is added in <https://vuepress-theme-hope.github.io/contribution/>, may take a while to deploy.

**Reply by dimaslanjaka** [reply]
> 1. we are using commonjs(cjs) on node side, and es module(esm) on client side, so you can not mixed it.

why not using pure tsconfig in root ?.
this my project [https://github.com/dimaslanjaka/universal-framework](https://github.com/dimaslanjaka/universal-framework) look at tsconfig.json, libs folder, and view folder.
- `libs/js` will deploy on browser side
- `libs/src` will used on node side
all functions and classes are from `libs/js` and export into `libs/src/compiler/framework.js` to be used on node side.
and also not produce `red underline` with eslint.
and still can be compiled with tsc.
and compatible with sublime, vscode, intellij idea, atom IDE typehinter/peek definitions/go to declaration/etc.

`vuepress-theme-hope` has aliases module, it cant resolving the source code while developing.
iam still confused, how to integrate `vuepress-theme-hope` into my IDE.

**Reply by dimaslanjaka** [reply]
> iam still confused, how to integrate `vuepress-theme-hope` into my IDE.

when i open one file from `packages` folder, it showing error. seems like package not found, module not found, etc.

**Reply by dimaslanjaka** [reply]
when i trying modify or paste some code or variables or constants. the file will shown error again because the IDE automated calculate file path from root not from the source copied file codes

**Reply by dimaslanjaka** [reply]
and when i open single folder from `packages` it shown errors. 
- Source module not found/not froom root path
- Aliased module not found
- when import from outer folder, will error too
- and so on

why not compiling them from root without bootstraping them on lerna ?. 
thats made development easily.

**Reply by dimaslanjaka** [reply]
> You should run ts:watch and copy:watch when you develop.

i already trying this, still producing errors on ide, and make typehint not defining source method/class/constants(global).

**Reply by dimaslanjaka** [reply]
btw, iam use laptop with 8gb ram and still HDD drive type.
i think, this project need SSD and better cpu and processor.
my ide cant follow the file changes on `vuepress-theme-hope` project.
iam sorry, i cant develop this project.

**Reply by dimaslanjaka** [reply]
but, iam still waiting next version `theme hope`.
i've deploy it https://github.com/dimaslanjaka/dimaslanjaka.github.io/tree/gh-pages -> https://dimaslanjaka.github.io

**Reply by Mister-Hope** [reply]
Everything works fine for me with VSCODE, and I also recommend you to try it if you have only 8GB ram.

No red lines will be showned if you open the entire forlder and if your ide is actually reading all files.

Every alia has a corresponding declaration file.

On vscode, you may need to reopen it after installation and a successful build, (otherwise you should restart ts server using command)

**Reply by Mister-Hope** [reply]
I am welcome to help if you are trying with VSCode. It's also the only recommended one mentioned in Vue3 official docs.

BTW, remember to install Vetur extension. It's required.

If you are still getting errors using VSCode, screenshot amd show me your errors.

**Reply by Mister-Hope** [reply]
> > 1. we are using commonjs(cjs) on node side, and es module(esm) on client side, so you can not mixed it.
> 
> why not using pure tsconfig in root ?.
> this my project [https://github.com/dimaslanjaka/universal-framework](https://github.com/dimaslanjaka/universal-framework) look at tsconfig.json, libs folder, and view folder.
> - `libs/js` will deploy on browser side
> - `libs/src` will used on node side
> all functions and classes are from `libs/js` and export into `libs/src/compiler/framework.js` to be used on node side.
> and also not produce `red underline` with eslint.
> and still can be compiled with tsc.
> and compatible with sublime, vscode, intellij idea, atom IDE typehinter/peek definitions/go to declaration/etc.
> 
> `vuepress-theme-hope` has aliases module, it cant resolving the source code while developing.
> iam still confused, how to integrate `vuepress-theme-hope` into my IDE.

For this, I strongly recommend you to look at official docs and understand what's composite mode before bargain with me. The composite mode is actually designed and recommended by typescript team to work under monorepos with different build requirement among projects.

**Reply by dimaslanjaka** [reply]
i'll try my best.
this seo need optimized.
default image when no image in article (fallback image).

![image](https://user-images.githubusercontent.com/12471057/121099068-604bfa00-c821-11eb-8184-d881030bbfc0.png)
![image](https://user-images.githubusercontent.com/12471057/121099206-98ebd380-c821-11eb-81c1-b13bf447b85b.png)

**Reply by dimaslanjaka** [reply]
Error components too

![image](https://user-images.githubusercontent.com/12471057/121099389-f2540280-c821-11eb-869b-b65bd5ac9c9b.png)

**Reply by Mister-Hope** [reply]
> Error components too
> 
> ![image](https://user-images.githubusercontent.com/12471057/121099389-f2540280-c821-11eb-869b-b65bd5ac9c9b.png)
> 

The demo is working as expected both on GitHub action and my users projects, check your code.

**Reply by Mister-Hope** [reply]
> i'll try my best.
> this seo need optimized.
> default image when no image in article (fallback image).
> 
> ![image](https://user-images.githubusercontent.com/12471057/121099068-604bfa00-c821-11eb-8184-d881030bbfc0.png)
> ![image](https://user-images.githubusercontent.com/12471057/121099206-98ebd380-c821-11eb-81c1-b13bf447b85b.png)
> 
> 

As you can see, the image is optional. So there is actually no issues.

 For all search engine, they won't display images in search results when the page doesn't have images, and that should be EXPECTED! Why should user find a unrelated default image in search result?

Additional: There is only one exception, which is the homepage,favicon will be used in some search engines.

You are warned, I won't accept this pr because it makes no sense.

**Reply by dimaslanjaka** [reply]
Breadcrumbs error
![image](https://user-images.githubusercontent.com/12471057/121100620-6e4f4a00-c824-11eb-891c-b0aa044f4613.png)

**Reply by dimaslanjaka** [reply]
> > i'll try my best.
> > this seo need optimized.
> > default image when no image in article (fallback image).
> > ![image](https://user-images.githubusercontent.com/12471057/121099068-604bfa00-c821-11eb-8184-d881030bbfc0.png)
> > ![image](https://user-images.githubusercontent.com/12471057/121099206-98ebd380-c821-11eb-81c1-b13bf447b85b.png)
> 
> As you can see, the image is optional. So there is actually no issues.
> 
> For all search engine, they won't display images in search results when the page doesn't have images, and that should be EXPECTED! Why should user find a unrelated default image in search result?
> 
> Additional: There is only one exception, which is the homepage,favicon will be used in some search engines.
> 
> You are warned, I won't accept this pr because it makes no sense.

okay

**Reply by Mister-Hope** [reply]
> Breadcrumbs error
> ![image](https://user-images.githubusercontent.com/12471057/121100620-6e4f4a00-c824-11eb-891c-b0aa044f4613.png)

Check if you have a valid title in every page. Please check carefully before reporting issues, thanks.

Besides, fell free ( and I recommand) to open a new discussion on each idea or bugs you would like to share. Nesting in a comment in dicussion can be hard to navigate and hard to track.

**Reply by dimaslanjaka** [reply]
> Check if you have a valid title in every page. Please check carefully before reporting issues, thanks.
> 
> Besides, fell free ( and I recommand) to open a new discussion on each idea or bugs you would like to share. Nesting in a comment in dicussion can be hard to navigate and hard to track.

This error because empty name.
![image](https://user-images.githubusercontent.com/12471057/121109612-20dbd880-c836-11eb-93f6-b7c95ae4552a.png)
Must be
![image](https://user-images.githubusercontent.com/12471057/121109700-45d04b80-c836-11eb-8d54-69f04655827c.png)
Finally
![image](https://user-images.githubusercontent.com/12471057/121109737-541e6780-c836-11eb-8db8-20981ff98114.png)

I have try bellow but still not fixed:
![image](https://user-images.githubusercontent.com/12471057/121109764-60a2c000-c836-11eb-882b-0fe16a200508.png)
Console log not working anymore 
![image](https://user-images.githubusercontent.com/12471057/121109819-76b08080-c836-11eb-877a-2ac16e5ae862.png)
no console message
![image](https://user-images.githubusercontent.com/12471057/121109863-8cbe4100-c836-11eb-8133-68b5376fd543.png) 
![image](https://user-images.githubusercontent.com/12471057/121109830-7dd78e80-c836-11eb-8c2c-1d73da32da77.png)

**Reply by dimaslanjaka** [reply]
how to develop when no debug ?

**Reply by Mister-Hope** [reply]
> > Check if you have a valid title in every page. Please check carefully before reporting issues, thanks.
> > 
> > Besides, fell free ( and I recommand) to open a new discussion on each idea or bugs you would like to share. Nesting in a comment in dicussion can be hard to navigate and hard to track.
> 
> This error because empty name.
> ![image](https://user-images.githubusercontent.com/12471057/121109612-20dbd880-c836-11eb-93f6-b7c95ae4552a.png)
> Must be
> ![image](https://user-images.githubusercontent.com/12471057/121109700-45d04b80-c836-11eb-8d54-69f04655827c.png)
> Finally
> ![image](https://user-images.githubusercontent.com/12471057/121109737-541e6780-c836-11eb-8db8-20981ff98114.png)
> 
> I have try bellow but still not fixed:
> ![image](https://user-images.githubusercontent.com/12471057/121109764-60a2c000-c836-11eb-882b-0fe16a200508.png)
> Console log not working anymore 
> ![image](https://user-images.githubusercontent.com/12471057/121109819-76b08080-c836-11eb-877a-2ac16e5ae862.png)
> no console message
> ![image](https://user-images.githubusercontent.com/12471057/121109863-8cbe4100-c836-11eb-8133-68b5376fd543.png) 
> ![image](https://user-images.githubusercontent.com/12471057/121109830-7dd78e80-c836-11eb-8c2c-1d73da32da77.png)
> 

like I said, set a title for every page, you should set title in frontmatter of you home page,just like my demo and template are doing.

**Reply by Mister-Hope** [reply]
Setting home as default makes no sense because there are situations where people forget to add a title in a simple page, and in that case,the breadcrumb should break (omit by search engines) rather then let search engines read a incorrect one, e.g: `bar/home/foo`

**Reply by Mister-Hope** [reply]
> how to develop when no debug ?

If you mean how to debug an output code, please open a new dicussion under Q&A category. It's not related to v2, and the conversation list is long enough scrolling down each time. Opening a new dicussion each time makes it easier to track.

**Reply by Mister-Hope** [reply]
Additional: if you insist, you can open a PR setting the breadcrumb default to `'Unknown'`, That's okey to me. @dimaslanjaka

**Reply by Mister-Hope** [reply]
Just to be clear, here is what you are missing:

https://github.com/vuepress-theme-hope/vuepress-theme-hope/blame/2b2e40eea9a920b88685edf2f9b7a0278e428c58/docs/theme/src/readme.md#L4

**Reply by Mister-Hope** [reply]
For debugging, make sure you are serving a project using local theme, e.g: demo or docs.

Besides, you should run `yarn dev` under root folder. And node side changes requires you to stop current devServer and start a new one.
